### PR TITLE
[WIP] Ignore JS's confirm, prompt and alert from background windows/tabs

### LINKED ISSUE #4106: Ignore window.confirm/alert/prompt in background tabs/windows
Other browsers do that as well, e.g. Chromium

- [window.confirm()](https://www.chromestatus.com/feature/5140698722467840) (Chrome 69)
- [window.alert()](https://www.chromestatus.com/feature/6477774290157568) (Chrome 64)
- [window.prompt()](https://www.chromestatus.com/feature/5637107137642496) (Chrome 56)

Also see https://docs.google.com/document/d/1wtV5rmLhbf1OZkbg7crtCt6h1mMtig_ctTQt3BLLEIU/edit

### PR DESCRIPTION
**Pull request by lufte**
Fixes #4106 

A couple of concerns:
1. Webengine seems to already be doing this for background tabs. It will show the dialog once you focus the tab. It doesn't do it for background windows though.
2. Personally, I think that the weird behavior is that a dialog from an inactive window/tab will show up in *every* window/tab. Perhaps if the dialog is only shown in its corresponding tab then we don't this fix, or is this behavior by design?
3. I'm always unsure on how to navigate the tree of widgets, is it okay to use `.parent()`? Please request every change you consider appropriate and I'll write tests afterwards.

### CODE REVIEW

File: qutebrowser/browser/browsertab.py [hunk 1]
```diff
@@ -1136,3 +1136,9 @@ def __repr__(self) -> str:
     def is_deleted(self) -> bool:
         assert self._widget is not None
         return sip.isdeleted(self._widget)
+
+    def is_in_background(self) -> bool:
+        active_tab = self.parent().currentWidget()
+        active_window = QApplication.instance().activeWindow()
+        return (self.tab_id != active_tab.tab_id
+                or self.win_id != active_window.win_id)
```
Comment (lufte) [inline qutebrowser/browser/browsertab.py]: `tab_id` seems to be global so I probably don't need to check for the win_id, right?

**Comment by The-Compiler** [comment]
> * Webengine seems to already be doing this for background tabs. It will show the dialog once you focus the tab. It doesn't do it for background windows though.

Could you try with older QtWebEngine versions as well?



> * Personally, I think that the weird behavior is that a dialog from an inactive window/tab will show up in _every_ window/tab. Perhaps if the dialog is only shown in its corresponding tab then we don't this fix, or is this behavior by design?

That changed at some point in qutebrowser's history, as for some prompts, we don't know which window/tab it's coming from - and the prompt code is already quite complex (because some prompts need to be blocking with their own event loop, and some can be asynchronous). I agree it's a bit weird, but changing it likely would lead to very hard to understand code.

> 3\. I'm always unsure on how to navigate the tree of widgets, is it okay to use `.parent()`? Please request every change you consider appropriate and I'll write tests afterwards.

That seems quite brittle and hard to follow - I haven't taken a closer look yet, but couldn't you just use something like `webview.window().isActive()` and `tab.isVisible()`?

**Comment by lufte** [comment]
> That seems quite brittle and hard to follow - I haven't taken a closer look yet, but couldn't you just use something like webview.window().isActive() and tab.isVisible()?

`tab.isVisible()` works, thanks, but I still need to get to the tab from the webview (or the `WebEnginePage` in the case of WebEngine) and I know you're trying to get rid of obgreg.

**Comment by The-Compiler** [comment]
Why do you need the tab? Can't you just call `isVisible()` on the QWeb(Engine)View?

**Comment by lufte** [comment]
Yes, sorry, I meant that I need to get to the webview from the webpage.

**Comment by The-Compiler** [comment]
You should be able to use `self.view()` for that, as there's `QWeb(Engine)Page::view()`.

**Comment by lufte** [comment]
Ah, I see. Well, you practically wrote this PR yourself :sweat_smile:

**Comment by The-Compiler** [comment]
oh, also it'd be good if you could add some tests for this to `tests/end2end/features/prompts.feature`.

**Comment by lufte** [comment]
> There's quite some code duplication there

I agree, but since I didn't find a common ancestor for `WebEnginePage` and `BrowserPage` where to put the common method (and I also don't see any use of multiple inheritance on the codebase) I decided to duplicate it. I'll put it in `shared.py`.
>oh, also it'd be good if you could add some tests for this to `tests/end2end/features/prompts.feature`.

Sure.

File: qutebrowser/browser/shared.py [hunk 2]
```diff
@@ -338,3 +338,8 @@ def netrc_authentication(url, authenticator):
     authenticator.setPassword(password)
 
     return True
+
+
+def is_js_dialog_allowed(page):
+    """Return if a webpage is allowed to raise an alert, prompt or confirm."""
+    return page.view().isVisible() and page.view().window().isActiveWindow()
```
Comment (lufte) [inline qutebrowser/browser/shared.py]: I decided not to log here to avoid passing the dialog type as a string (when it should be just 3 valid values) and to keep this a pure `is_X` boolean function.
